### PR TITLE
Fix the glob_ext() filesystem globbing function

### PR DESCRIPTION
**Pull request by pyrco**
When passed a pattern with path parts which did not contain any globs, the function would fail due to path strings instead of FilesystemEntry's being used.

Documentation and comments are added to the glob_ext*() functions and they are refactored a bit.

(DIS-2903)

### CODE REVIEW

File: dissect/target/helpers/fsutil.py [hunk 1]
```diff
@@ -857,7 +857,18 @@ def walk_ext(path_entry, topdown=True, onerror=None, followlinks=False):
         yield [path_entry], dirs, files
 
 
-def glob_split(pattern: str, alt_separator: str = "") -> str:
+def glob_split(pattern: str, alt_separator: str = "") -> tuple[str, str]:
+    """
+    Split a pattern on path part boundaries on the first path part with a glob pattern.
```
Comment (Schamper) [inline dissect/target/helpers/fsutil.py:862]:
```suggestion
    """Split a pattern on path part boundaries on the first path part with a glob pattern.
```

File: dissect/target/helpers/fsutil.py [hunk 2]
```diff
@@ -870,75 +881,118 @@ def glob_split(pattern: str, alt_separator: str = "") -> str:
     return pattern[:pos], pattern[pos:]
 
 
-def glob_ext(direntry: filesystem.FilesystemEntry, pattern: str) -> filesystem.FilesystemEntry:
+def glob_ext(direntry: filesystem.FilesystemEntry, pattern: str) -> Iterator[filesystem.FilesystemEntry]:
+    """
+    Recursively search and return filesystem entries matching a given glob pattern.
```
Comment (Schamper) [inline dissect/target/helpers/fsutil.py:886]:
```suggestion
    """Recursively search and return filesystem entries matching a given glob pattern.
```

File: dissect/target/helpers/fsutil.py [hunk 3]
```diff
@@ -870,75 +881,118 @@ def glob_split(pattern: str, alt_separator: str = "") -> str:
...
+        for entry in glob_in_base(direntry, base_name):
             yield entry
 
 
 # These 2 helper functions non-recursively glob inside a literal directory.
-# They return a list of basenames. `glob1` accepts a pattern while `glob0`
-# takes a literal base_name (so it only has to check for its existence).
+def glob_ext1(direntry: filesystem.FilesystemEntry, pattern: str) -> Iterator[filesystem.FilesystemEntry]:
+    """
+    Match and return filesystem entries in a given filesystem entry based on pattern.
```
Comment (Schamper) [inline dissect/target/helpers/fsutil.py:948]:
```suggestion
    """Match and return filesystem entries in a given filesystem entry based on pattern.
```

File: dissect/target/helpers/fsutil.py [hunk 4]
```diff
@@ -870,75 +881,118 @@ def glob_split(pattern: str, alt_separator: str = "") -> str:
...
-            yield e
+            yield entry
 
 
-def glob_ext0(direntry: filesystem.FilesystemEntry, base_name: str) -> list[filesystem.FilesystemEntry]:
-    if base_name == "":
-        # `os.path.split()` returns an empty base_name for paths ending with a
+def glob_ext0(direntry: filesystem.FilesystemEntry, path: str) -> Iterator[filesystem.FilesystemEntry]:
+    """
+    Return the filesystem entry equal to the given path relative to direntry.
```
Comment (Schamper) [inline dissect/target/helpers/fsutil.py:976]:
```suggestion
    """Return the filesystem entry equal to the given path relative to ``direntry``.
```

File: dissect/target/helpers/fsutil.py [hunk 5]
```diff
@@ -870,75 +881,118 @@ def glob_split(pattern: str, alt_separator: str = "") -> str:
...
+
+    Args:
+        direntry: The filesystem entry relative to which to return the path.
+        path: The path to the filesystem entry to return (if present). If path
+              is the empty string (``''``) the ``direntry`` itself is returned.
```
Comment (Schamper) [inline dissect/target/helpers/fsutil.py:981]:
```suggestion
              is an empty string (``''``) the ``direntry`` itself is returned.
```

File: dissect/target/helpers/fsutil.py [hunk 6]
```diff
@@ -870,75 +881,118 @@ def glob_split(pattern: str, alt_separator: str = "") -> str:
...
     if not dir_name:
         for entry in glob_ext1(direntry, base_name):
             yield entry
         return
 
-    if dir_name != pattern and has_glob_magic(dir_name):
-        dirs = glob_ext(direntry, dir_name)
+    # If the pattern has more than one path part and these parts (dir_name)
+    # contain globs, we recursively go over all the path parts and match them
+    # (glob_ext).
```
Comment (Schamper) [inline dissect/target/helpers/fsutil.py:924]:
```suggestion
    # If the pattern has more than one path part and these parts (dir_name) contain globs,
    # we recursively go over all the path parts and match them (glob_ext).
```

File: dissect/target/helpers/fsutil.py [hunk 7]
```diff
@@ -870,75 +881,118 @@ def glob_split(pattern: str, alt_separator: str = "") -> str:
...
+    # If these path parts have no globs, we get the path directly by name
+    # (glob_ext0).
```
Comment (Schamper) [inline dissect/target/helpers/fsutil.py:926]:
```suggestion
    # If these path parts have no globs, we get the path directly by name (glob_ext0).
```

File: dissect/target/helpers/fsutil.py [hunk 8]
```diff
@@ -870,75 +881,118 @@ def glob_split(pattern: str, alt_separator: str = "") -> str:
...
+            # match only directories.
             if base_name:
                 yield entry
-            # Patterns ending with a slash should match only directories
             elif entry.is_dir():
                 yield entry
         return
 
+    # The pattern has only one path part, so we can match directly against the
+    # files in direntry.
```
Comment (Schamper) [inline dissect/target/helpers/fsutil.py:916]:
```suggestion
    # The pattern has only one path part, so we can match directly against the files in direntry.
```

File: dissect/target/helpers/fsutil.py [hunk 9]
```diff
@@ -870,75 +881,118 @@ def glob_split(pattern: str, alt_separator: str = "") -> str:
...
 
+    # The simple case where there are no globs.
     if not has_glob_magic(pattern):
         try:
             entry = direntry.get(pattern)
         except FileNotFoundError:
             pass
         else:
+            # Patterns ending with a slash, so without a base_name, should
+            # match only directories.
```
Comment (Schamper) [inline dissect/target/helpers/fsutil.py:908]:
```suggestion
            # Patterns ending with a slash, so without a base_name, should match only directories.
```

**Review by Schamper — CHANGES_REQUESTED** [review]
Maybe also add a unit test for the previously failing case?

**Comment by Schamper** [comment]
No regression unit test? :smile:

**Comment by pyrco** [comment]
I added some tests :wink:

**Review by Schamper — APPROVED** [review]

**Pull request merged**